### PR TITLE
V3—point-selection

### PR DESCRIPTION
**Pull request by bfinzer**
Clicking, dragging, animating and hovering logic for points in graphs is very subtle. This pull request goes quite a ways toward unifying and reusing code.

### CODE REVIEW

File: v3/src/components/graph/components/background.tsx [hunk 1]
```diff
@@ -56,19 +56,20 @@ export const Background = forwardRef<SVGGElement, IProps>((props, ref) => {
     height = useRef(0),
     selectionTree = useRef<typeof RTree | null>(null),
     previousMarqueeRect = useRef<rTreeRect>(),
-    currentlySelectedCaseIDs = useRef<string[]>([]),
 
-    onDragStart = useCallback((event: MouseEvent) => {
+    onDragStart = useCallback((event:any) => {
```
Comment (kswenson) [inline v3/src/components/graph/components/background.tsx:60]: Seems like there must be a better way of typing this.

File: v3/src/components/graph/components/chartdots.tsx [hunk 2]
```diff
@@ -124,9 +125,22 @@ export const ChartDots = memo(function ChartDots(props: IProps) {
       baseCoord = primaryIsBottom ? 0 : layout.axisLength('left'),
       signForOffset = primaryIsBottom ? 1 : -1,
       primaryCenterKey = primaryIsBottom ? 'cx' : 'cy',
-      secondaryCenterKey = primaryIsBottom ? 'cy' : 'cx'
+      secondaryCenterKey = primaryIsBottom ? 'cy' : 'cx',
 
-    const setPoints = () => {
+      lookupLegendColor = (id: string) => {
+        const isSelected = dataset?.isCaseSelected(id),
+          legendColor = getLegendColor ? getLegendColor(id) : ''
```
Comment (kswenson) [inline v3/src/components/graph/components/chartdots.tsx:132]:
```suggestion
          legendColor = getLegendColor?.(id) ?? ''
```

File: v3/src/components/graph/hooks/graph-hooks.ts [hunk 3]
```diff
@@ -136,15 +137,65 @@ export const usePlotResponders = (props: IPlotResponderProps) => {
...
 }
+
+const dataTip = d3tip().attr('class', 'graph-d3-tip')/*.attr('opacity', 0.8)*/
+  .attr('data-testid', 'graph-point-data-tip')
+  .html((d: string) => {
+    return "<p>" + d + "</p>"
+  })
+
+export const useDataTips = (dotsRef: React.RefObject<SVGSVGElement>,
+                            dataset: IDataSet | undefined, graphModel: IGraphModel) => {
```
Comment (kswenson) [inline v3/src/components/graph/hooks/graph-hooks.ts:161]: This module is getting pretty long. Seems like `useDataTips` merits its own `use-data-tips.ts` module.

File: v3/src/components/graph/utilities/graph-utils.ts [hunk 4]
```diff
@@ -101,15 +101,38 @@ export function getPointTipText(caseID: string, attributeIDs: string[], dataset?
...
   caseIDs: string[]
   dotsElement: SVGGElement | null
   pointRadius: number
   enableAnimation: React.MutableRefObject<boolean>
   instanceId: string | undefined
 }
 
+export function handleClickOnDot(event: MouseEvent, caseID: string, dataset?: IDataSet) {
+  const [, caseId] = caseID.split("_"),
+    extendSelection = event.shiftKey,
```
Comment (kswenson) [inline v3/src/components/graph/utilities/graph-utils.ts:114]:
```suggestion
    extendSelection = event.shiftKey || event.metaKey,
```
Cmd/meta key should behave the same way.

**Review by kswenson — CHANGES_REQUESTED** [review]
👍 The refactor and reuse here look great -- just a few small suggestions.

File: v3/src/components/graph/hooks/graph-hooks.ts (hunk 3 above)
Comment (bfinzer) [reply v3/src/components/graph/hooks/graph-hooks.ts:161]: Done

File: v3/src/components/graph/components/background.tsx (hunk 1 above)
Comment (bfinzer) [reply v3/src/components/graph/components/background.tsx:60]: Yes, there was.

**Comment by bfinzer** [comment]
But Shift is not Cmd, so I don't think metKey applies

**Comment by kswenson** [comment]
> But Shift is not Cmd, so I don't think metKey applies

UI guidelines generally treat cmd/meta as discontiguous selection while shift is contiguous selection. In situations where there is no notion of contiguous selection, as is true for the graph, they should behave the same.

**Pull request merged**